### PR TITLE
Fix prometheus format of metrics.

### PR DESCRIPTION
**Pull request by loburm**
Two major updates:
1. Metrics are grouped, and each group has TYPE annotation.
2. Added process_start_time_seconds metric.

### CODE REVIEW

File: src/http_server/api/v1/metrics.c [hunk 1]
```diff
@@ -139,6 +139,38 @@ static void cb_mq_metrics(mk_mq_t *queue, void *data, size_t size)
...
 }
 
+int string_cmp(const void* a_arg, const void* b_arg) {
+  char* a = *(char **)a_arg;
+  char* b = *(char **)b_arg;
+  return strcmp(a, b);
+}
+
+size_t extract_metric_name_end_position(char *s) {
+    int len = strlen(s);
```
Comment (edsiper) [inline src/http_server/api/v1/metrics.c]: this is not necessary. If you look from the caller, what you are really passing as a reference is a flb_sds_t string, so you can use flb_sds_len(s) and avoid a strlen().

File: src/http_server/api/v1/metrics.c [hunk 2]
```diff
@@ -180,14 +220,30 @@ void cb_metrics_prometheus(mk_request_t *request, void *data)
...
+        msgpack_object v = map.via.map.ptr[i].val;
+        /* Iterate sub-map */
+        for (j = 0; j < v.via.map.size; j++) {
+            msgpack_object sv = v.via.map.ptr[j].val;
+            for (m = 0; m < sv.via.map.size; m++) {
+                num_metrics++;
+            }
+        }
+    }
+    metrics_arr = malloc(num_metrics * sizeof(char*));
```
Comment (edsiper) [inline src/http_server/api/v1/metrics.c]: here you need to use our wrapper flb_malloc()

File: src/http_server/api/v1/metrics.c [hunk 3]
```diff
@@ -211,23 +267,48 @@ void cb_metrics_prometheus(mk_request_t *request, void *data)
...
                 mv = sv.via.map.ptr[m].val;
-
-                sds = flb_sds_cat(sds, "fluentbit_", 10);
-                sds = flb_sds_cat(sds, k.via.str.ptr, k.via.str.size);
-                sds = flb_sds_cat(sds, "_", 1);
-                sds = flb_sds_cat(sds, mk.via.str.ptr, mk.via.str.size);
-                sds = flb_sds_cat(sds, "_total{name=\"", 13);
-                sds = flb_sds_cat(sds, sk.via.str.ptr, sk.via.str.size);
-                sds = flb_sds_cat(sds, "\"} ", 3);
+                sds_metric = flb_sds_create_size(1024);
```
Comment (edsiper) [inline src/http_server/api/v1/metrics.c]: note that sds_metric is never freed at the end of the function, so it's leaking memory

File: src/http_server/api/v1/metrics.c [hunk 4]
```diff
@@ -211,23 +267,48 @@ void cb_metrics_prometheus(mk_request_t *request, void *data)
...
+                sds_metric = flb_sds_cat(sds_metric, tmp, len);
+                sds_metric = flb_sds_cat(sds_metric, time_str, time_len);
+                sds_metric = flb_sds_cat(sds_metric, "\n", 1);
+                metrics_arr[index] = sds_metric;
+                index++;
             }
         }
     }
+
+    // Sort metrics in alphabetic order, so we can group them later.
```
Comment (edsiper) [inline src/http_server/api/v1/metrics.c]: please use /* comment */ style as described in our contribution guidelines:

https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md

File: src/http_server/api/v1/metrics.c [hunk 5]
```diff
@@ -237,6 +318,11 @@ void cb_metrics_prometheus(mk_request_t *request, void *data)
                    PROMETHEUS_HEADER, sizeof(PROMETHEUS_HEADER) - 1);
     mk_http_send(request, sds, flb_sds_len(sds), NULL);
     mk_http_done(request);
+
```
Comment (edsiper) [inline src/http_server/api/v1/metrics.c]: the following block must be always placed _before_ the call mk_http_done(), there are some cases where mk_http_done() can return right away (it's a coroutine and might yield)

File: src/http_server/api/v1/metrics.c (hunk 1 above)
Comment (loburm) [reply src/http_server/api/v1/metrics.c]: Thanks, done.

**Comment by loburm** [comment]
Thank you for the comments!
I have resolved all of them, please take another look.

**Comment by loburm** [comment]
@edsiper can you take a look on this PR?

**Comment by loburm** [comment]
@edsiper ping

**Comment by edsiper** [comment]
thanks

**Pull request merged**